### ISSUE TITLE
Add support for asynchronously awaiting nothing to pause execution for one cycle.

### ISSUE THREAD

**Issue by mathgeniuszach** [godotengine/godot-proposals#10943]
### Describe the project you are working on

A lisp-like code interpreter to use with a generic PEG parser (with syntax similar to [peggyjs](https://peggyjs.org/documentation.html#grammar-syntax-and-semantics)) and sandboxing library.

Some example code I'm using for testing at the moment:
```gdscript
func _ready():
	var runner := PEGRunner.new({
		"+": [[true, true], (func(r, args): return args[0] + args[1])],
		"*": [[true, true], (func(r, args): return args[0] * args[1])],
		"print": [[true], func(r, args): print(args[0])],
		"if": [[true, false, false], (func(r, args): return PEGRunner.evaluate(args[1] if args[0] else args[2]))],
		"run": [false, (func(r, args): r.push_instrs(args))],
		"while": [[false, false], (func(r, args): r.push_loop([args[1]], args[0]))]
	})
	runner.run([
		["while", 1, ["print", "loop"]],
		["print", "exit"]
	])
```

### Describe the problem or limitation you are having in your project

In this project, the code interpreter is able to run just one instruction at a time (and then pause execution), thus it is possible to run many interpreters at once. To streamline the process of doing so, I wanted to provide access to an asynchronous function, `run_async()`, which runs a single instruction, then pauses execution for another program or item to take over.

As `await 0` or similar doesn't actually await anything (unlike other languages like javascript, which will pause execution for another function to take over if available), the only way I can force execution to pause for a minimal amount of time is with `await get_tree().create_timer(0).timeout`, but I'm not sure if this is creating unnecessary slowdowns (I don't know how long it takes to create a timer and for the signal to come through, or if it's even faster than get_tree().process_frame).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Adding the ability to await nothing gives a convenient way for long-running asynchronous functions to signal a safe place to stop and let something else work if necessary. While awaiting get_tree().create_timer(0).timeout or get_tree().process_frame could create extra overhead or slowdown the program, awaiting nothing signals, very clearly, to just pause for one cycle.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Consider the following javascript code:
```javascript
async function f(t) {
    for (let i = 0; i < 10; i++) {
        await 0; console.log(t+String(i))
    }
}
f("a")
f("b")
```
This prints (not necessarily in order, but not synchronously):
```
a0
b0
a1
b1
...
a9
b9
```
I would like the following GDScript code to function the same way:
```gdscript
func f(t):
	for i in 10:
		await # or "await 0"
		print(t+str(i))

func _ready():
	f("a")
	f("b")
```
so it defers the asynchronous code at least once, without needing to use get_tree() either. Currently, it is possible to do "await 0" in the above code, but this runs synchronously and doesn't pause the execution cycle.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Like I said before, it _is_ possible to achieve something similar with `await get_tree().create_timer(0).timeout` or `await get_tree().process_frame`, but these methods take additional overhead and can unnecessarily slow down program execution. Having this as a language feature would enable a faster way of writing long-running asynchronous functions.

### Is there a reason why this should be core and not an add-on in the asset library?

This is a language implementation/syntax feature, so it cannot be implemented in an external library.

**Comment by dalexeev** [comment]
This won't work as you expect. `await` waits for either a signal to be emitted or a coroutine to complete (i.e. a suspended function call that is waiting for either a signal to be emitted or a coroutine to complete). If the `await` operand is neither a signal nor a coroutine, `await` simply returns that value immediately (this is to solve the problems of `yield` from 3.x).

What you propose still requires defining some specific moment at which execution should continue. Coroutines themselves are not related to threads, we can't "just" execute asynchronous code in parallel. In the case of `create_timer(0)`, the `timeout` signal will be emitted at the frame end, when `SceneTree` checks the remaining time on the timers.

JavaScript seems to use a similar mechanism:

[await - Control flow effects of await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#control_flow_effects_of_await)

> When an await is encountered in code (either in an async function or in a module), the awaited expression is executed, while all code that depends on the expression's value is paused and pushed into the [microtask queue](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Event_loop). The main thread is then freed for the next task in the event loop. This happens even if the awaited value is an already-resolved promise or not a promise.

Yes, GDScript behavior is different from JavaScript regarding the last sentence. We do not suspend the call if the `await` operand is not a "promise" (signal or coroutine). But also in GDScript you can't mark a function as `async`, and even having `await` in the function body doesn't make it asynchronous. This is determined at runtime depending on whether execution has reached `await` and what its operand is.

[The event loop - Zero delays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Event_loop#zero_delays)

> Zero delay doesn't mean the call back will fire-off after zero milliseconds. Calling [setTimeout()](https://developer.mozilla.org/en-US/docs/Web/API/Window/setTimeout) with a delay of 0 (zero) milliseconds doesn't execute the callback function after the given interval.

In Godot, the event loop is available in a more explicit form (`SceneTree` or `MainLoop`, `call_deferred()`, etc.), so it makes sense that interacting with it also often requires explicit actions from the user.

The only other way I can think of is to wait until the parent call leaves the call stack (or the entire call stack is cleared?), but that seems like a confusing and probably inefficient solution.

**Comment by mathgeniuszach** [comment]
I think I'm a little confused as to what "await" actually does, then. What occurs under the hood the moment an "await" is reached and it's operand is a signal or coroutine? My assumption was that it caches the function as a coroutine, pushes it on the bottom of the event loop, and then returns back to the main event loop for another item to be called in the loop. After each item is completed in the event loop, it moves on to the next, until the coroutine is reached again, at which it either continues the cached function, or if the value has not been resolved yet, pushes it back onto the bottom of the event loop. This is the conventional way in which most asynchronous event loops work, but is this not true of Godot?

Looking into call_deferred(), I am a little confused on this statement:
> Deferred calls are processed at idle time. Idle time happens mainly at the end of process and physics frames. In it, deferred calls will be run until there are none left, which means you can defer calls from other deferred calls and they'll still be run in the current idle time cycle. This means you should not call a method deferred from itself (or from a method called by it), as this causes infinite recursion the same way as if you had called the method directly.

This seems to contradict the notion of a global event loop, in which process and physics frames are _also_ on the event loop, and "infinite recursion" (or freezing) cannot occur because they alternate execution with other elements on the loop (as items are pushed onto the bottom). Are these frames not on the event loop?

The idea of awaiting nothing or a regular value is to simply cache the function as a coroutine and push it onto the bottom of the event loop. When the cached function is reached again, it continues it regardless of awaited value.

**Comment by MajorMcDoom** [comment]
I know this is super late, but I thought I'd just comment in case it helps the proposal reach some kind of resolution.

@mathgeniuszach There is no global "event loop" for coroutines. Coroutines in Godot work very differently. When an `await x` is reached, it quite literally waits for whatever `x` is before it executes the next line. So for example, if you have:
```
await some_signal
do_some_stuff()
```

Then `do_some_stuff()` is literally executing in direct response to `some_signal`, _as though it was a connected callback_. It's effectively equivalent to this:
```
some_signal.connect(do_some_stuff)
```

In fact, if you try to do something like this, it would throw an error, because you're not allowed to free an object that is currently emitting a signal:
```
await enemy.damaged
enemy.free() # This will throw an error.
```

If you await a function call instead of a signal, then it instantly returns control after that entire function has been executed. It can be instant, or, if that function itself ends up awaiting other signals, then control is returned as though it were connected to the last signal that was awaited in that function. So for example, if you have:

```
await do_some_stuff()
print("B")
...
func do_some_stuff():
    await some_signal
    print("A")
```

This is effectively equivalent to:
```
some_signal.connect(func():
    print("A")
    print("B"))
```

TL;DR; you can think of coroutines as really just a chain of signal connections. At no point does it allow for "interruptions", or pass control over to the main loop.

As a solution to the specific problem you presented, you can simply give your runner its own public `step()` function, and make that function emit an internal `stepped` signal. Then all your coroutine code can simply be interleaved with `await stepped` whenever it needs to take a break. This allows you to manually execute as many or as few lines in instant succession.

**Comment by mathgeniuszach** [comment]
@MajorMcDoom My usecase was for sandboxing code that should run simultaneously with main code, so a step function is not viable. I needed the code to run for as much time is left during the frame, but without blocking other code. If I did a small number of instructions per frame, that would obviously be too slow. If I did too many, the frame would finish too late. There is no sweet spot - each frame will be different depending on other code.

I have a super hacky workaround that calls `await get_tree().create_timer(0).timeout` repeatedly normally until just before the end of a computed frame (I manually check the frame time), then await for a process frame. In any other language, I await nothing or an empty timer, this hacky workaround is not a real solution.

I want to give everything else a chance to process, but if there's more time in the frame I want to use it. But I don't want to indefinitely postpone the frame.

**Comment by MajorMcDoom** [comment]
> [@MajorMcDoom](https://github.com/MajorMcDoom) My usecase was for sandboxing code that should run simultaneously with main code, so a step function is not viable. I needed the code to run for as much time is left during the frame, but without blocking other code. If I did a small number of instructions per frame, that would obviously be too slow. If I did too many, the frame would finish too late. There is no sweet spot - each frame will be different depending on other code.
> 
> I have a super hacky workaround that calls `await get_tree().create_timer(0).timeout` repeatedly normally until just before the end of a computed frame (I manually check the frame time), then await for a process frame. In any other language, I await nothing or an empty timer, this hacky workaround is not a real solution.
> 
> I want to give everything else a chance to process, but if there's more time in the frame I want to use it. But I don't want to indefinitely postpone the frame.

I think you're misunderstanding what `create_timer(0).timeout` is actually doing. If you repeatedly await that, you're actually waiting one frame per await. Timers are updated once per frame, so when you receive the first `timeout` signal, awaiting another one would make you wait until the next timer update, which would be the next frame.

<img width="595" height="349" alt="Image" src="https://github.com/user-attachments/assets/c7ad96a9-74cf-42a3-8785-7b36c23d7e65" />

If your goal is to just keep a consistent frame rate, then just use the step method I previously mentioned, but just check the OS time after every step.

**Comment by MajorMcDoom** [comment]
Here's some sample code to show what I mean.
```gdscript
class_name Test
extends Node

signal stepped()

var x: int
var prev_x: int
var prev_ms: int

func _ready() -> void:

	my_coroutine()
	prev_ms = Time.get_ticks_msec()

func _process(delta: float) -> void:

	var ms: int
	while true:
		step()
		ms = Time.get_ticks_msec()
		var elapsed := (ms - prev_ms) * 0.001
		if elapsed >= 0.05:
			print("frame: %d, elapsed: %f, steps: %d" % [Engine.get_process_frames(), elapsed, x - prev_x])
			break
	prev_ms = ms
	prev_x = x

func my_coroutine() -> void:
	while true:
		x += 1
		await stepped

func step() -> void:
	stepped.emit()
```
Here's the first ten lines of output. Note that the elapsed milliseconds per frame is consistent, and the steps of coroutine performed per frame starts to stabilize at around ~19400. Obviously in your case, you'd have to do additional calculations to only take up _remaining_ time in a frame, but the concept is the same.
```
frame: 0, elapsed: 0.050000, steps: 16419
frame: 1, elapsed: 0.050000, steps: 7620
frame: 2, elapsed: 0.050000, steps: 19334
frame: 3, elapsed: 0.050000, steps: 19344
frame: 4, elapsed: 0.050000, steps: 19447
frame: 5, elapsed: 0.050000, steps: 19429
frame: 6, elapsed: 0.050000, steps: 19371
frame: 7, elapsed: 0.050000, steps: 19388
frame: 8, elapsed: 0.050000, steps: 19452
frame: 9, elapsed: 0.050000, steps: 19541
frame: 10, elapsed: 0.050000, steps: 19403
```
